---
title: "Prerequisites for 16S Data Processing Training"
author: Katie McCauley^[Bioinformatics and Computational Biology Branch (BCBB), OCICB, NIAID, NIH]
date: "`r Sys.Date()`"
output: html_document
---

We're excited you're joining us for the upcoming 16S rRNA data processing training! This document is intended to help set up your environment before joining the class so we can jump right in with analyzing data.

## Computer Setup

For now, we'll do most of the analysis on your personal computer in R and RStudio. You can download both without administrator privileges by visiting the NIH Self Service Portal.

In addition to installing the software we'll use, please also install a few packages that we'll use during the workshop by copying and pasting the lines below into the "Console" (the panel in the lower left when you open a fresh installation of Rstudio -- it should have a carrot sign `>`). The script can also be found at `install.R`.

```{r, comment=NA}
setwd("~/Documents/training/16S-data-processing/")
cat(readLines("install.R"), sep="\n")
```

## Bringing Your Own Data

We plan to provide you with data to use on your own computer for analyzing during the tutorial. However, if you would like to analyze your own data, you are more than welcome to bring some! If you choose to go this route, please be aware that your data will need to satisfy certain crteria.

First, you will need FASTQ-formatted data, which means raw sequence data with quality information. If you look at your files in a text editor (like TextEdit, Notepad++, vim, Sublime, etc), they may look something like this:

```
@M03213:59:000000000-AWR6D:1:1101:12406:1145 2:N:0:NCCTGAGC+NTATTAAG
NGACTACTGGGGTTTCTAATCCTGTTTGCTCCCCACGCTTTCGCACATGAGCGTCAGTACATTCCCAAGNGGCTGCCTTCGCCTTCGGTATTCCTCCACATCTCTACGCNTTTCACCGCTACACGTGGAATTCTACCCCTCCCTAAAGTACTCTAGATTCCCAGTCTGAAATGCAATTCCCAGGTTAAGCCCGGGGCTTTCACACCTCACTTAAAAATCCGCCTGCGTGCCCTTTACGCCCAGTTATTCCGATTAACGCT
+
#8ACCGGGGGGGFFGGGGFGGGGGGGGGGGGGGGGGGGGGGGEGGFFGGGGGGGGGGGGGGFGFFGG<E#:BFFGGGFGGGGGCGGFEFGFFGGGGG<CFFCFGGGGGG#99@FFGGEGBGGFGGF8CFFFEFGGG<=9DC>DDGGD?C=,;EGFGBFDGFFGGGGCC;@EEFGGGGFGGGGGGGGGFGGFDEGGGAADE5;EEB*07/9<FFCFGFGD@=@EDFF>7>9;C?E</2(2.6;<=)7,9<?(*)6(7,,(-
@M03213:59:000000000-AWR6D:1:1101:9817:1174 2:N:0:NCCTGAGC+CTATTAAG
NGACTACTGGGGTTTCTAATCCTGTTCGCTACCCACGCTTTCGAGCCTCAGCGTCAGTTACAAGCCAGAGAGCCGCTTTCGCCACAGGTGTTCCTCCATATATCTACGCATTTCACCGCTACACATGGAATTCCACTCTCCCCTCTTGCACTCAAGTTAAACAGTTTCCAAAGCAAACTATGGTTGAGCCACAGCCTTTGACTTCAGACTTATCTAACCGCCTGCGCTCGCTTTCCGCCCACTAAATCCGTATAACTCTCG
+
#8ACCGGGGGGGFCGGGGGGGGGGGGFGGGGGGGGGGGGGGGGFGGGGFGFFGGGFGFGGGGGGG7CFGCFFGGGGBEGGGGGGGG?EGGGGGFGGGGGGGGFGGGGGGGE@DFAFGGGFGEGGGGGGGGGF<,@,DDFGGDGDG=EFGGGFF,=D?8DF,?EGGCFCF,DFFGGFCDGFGG8@E3<?FF8DG8BFDFEEFCBEFA7;@6;A@CGGC7915>8)702/8:4*4A+7=;((/(,/6<29(((*.//,,/)/(
```

We will cover what all of these letters mean when you join us, but to simply confirm that you have "FASTQ" data, you want to look for the plus signs every few lines.

Second, these files need to be separated by sample ("demultiplexed"), so if you are analyzing 20 samples, you'll need to have 40 files. This also brings up another point -- that while you can bring single-end data to the tutorial, we will be focusing on paired-end data where there are forward and reverse reads that we need to merge together. I will try to point out when a modification for single-end data is needed, but will be unable to troubleshoot issues during the tutorial.

The sequence data also needs to be without non-biological sequences, including primers and adapters. You can check for the presence of these with a tool called [FASTQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/), and remove them with tools like [fastp](https://github.com/OpenGene/fastp), [trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic), or [cutadapt](https://cutadapt.readthedocs.io/en/stable/). Alternatively, you can submit your data to [Nephele](https://nephele.niaid.nih.gov) under the "Pre-process" tab, which uses several of these tools without the need for the command line^[https://nephele.niaid.nih.gov/details_qc/]. For now, there is no need to worry about quality trimming, and please do not merge paired-end reads -- we'll do both with DADA2. If you run into any issues, feel free to e-mail nephelesupport@nih.gov.

Finally, if you have a large dataset, consider only working on ~5-10 samples during the workshop. The methods we use should be flexible enough that you can apply the concepts to more samples after the workshop, and we're always available for questions! (bioinformatics@niaid.nih.gov).

## Resources

We plan to draw heavily from the (tutorial developed by Benjamin Callahan for DADA2)[https://benjjneb.github.io/dada2/tutorial.html], and encourage you to scan that tutorial prior to the training.